pkg/strategy/streaming.go

```diff
@@ -25,7 +27,7 @@ import (
 	"github.com/jaegertracing/jaeger-operator/pkg/util"
 )
 
-func newStreamingStrategy(ctx context.Context, jaeger *v1.Jaeger) S {
+func newStreamingStrategy(ctx context.Context, jaeger *v1.Jaeger, es *storage.ElasticsearchDeployment) S {
```

Comment:
This doesn't seem to belong here... We can't expect to add a `cassandra`, `badger`, `in-memory` to the strategies. There has to be a better way to do this.

Reply:
The main motivation is to set path to cert generation scrip. Tests use a different path than binary.

Reply:
Can this be done via viper, or added to the context? Sounds a bit outside of the scope to change the strategy signature just to be able to get some information internal to the auto-provisioning of one of the storages...

Reply:
So when creating the self-provisioned ES we need two things: path to cert generation script and more importantly secretes from the environment. 

The path could be done via viper (although there are downsides of using viper in the impl packages, it's cleaner to pass params directly if we are talking about a package down in the stack).

Anyways the production strategy is using exactly the same construction.

Reply:
So when creating the self-provisioned ES we need two things: path to cert generation script and more importantly secretes from the environment.

The path could be done via viper (although there are downsides of using viper in the impl packages, it's cleaner to pass params directly if we are talking about a package down in the stack).

Anyways the production strategy is using exactly the same construction.

Reply:
It sounds reasonable to me to require access to a configuration value (via viper or not) if the path to the cert generation is required. In fact, we are using viper already in this source for other purposes:

https://github.com/jaegertracing/jaeger-operator/blob/5999ea6f1057cc63d30888272e9f101a697da8f6/pkg/strategy/streaming.go#L90-L98

Reply:
Given the same approach is used in the production strategy I would like to keep it consistent and if required do refactoring in a separate PR.

Anyways could you please review other parts of the PR?

Reply:
That part was also contentious when it was added. I'd rather do it right this time and fix the production strategy later, than to have it wrong in both places.

Reply:
I am happy to hear any suggestions on how to get around this.

At the moment we create this object which is passed to prod and streaming strategy:

```
es := &storage.ElasticsearchDeployment{Jaeger: jaeger, CertScript: esCertGenerationScript, Secrets: secrets}
```

`es` object holds three params jaeger cr, path to cert generation script and secrets from the environment. Jaeger object could be removed because it's also passed to the strategy.
To properly handle ES secrets we need the path to cert generation and existing secrets.

To completely remove this object we would have to use viper for the cert generation scrip path and pass k8s client to the strategy (or create a new client, I am not sure if SKD provides API to a globally cached client).

> It sounds reasonable to me to require access to a configuration value (via viper or not) if the path to the cert generation is required. In fact, we are using viper already in this source for other purposes:

Using viper in high-level packages might be fine, although I do not like to depend on it in low-level components. These should have clear API contracts. Then it's easier to test and understand external package requirements.

Reply:
The cert path seems to be something that is installation-dependent, rather than CR-dependent, so, it's value belongs to a config store instead of parameter to method calls related to the reconciliation loop.

So, the only thing that needs a solution is about the secrets. I assume you need it to get access to the username/password to use for ES, right? Is it the case, that there can only be one provisioned ES per cluster? If so, you could create a secret with the username/pass in the format that the deployments could consume, like described in #846.

Reply:
> I assume you need it to get access to the username/password to use for ES, right?

No, we don't use user/pass in self-provisioned ES. The secrets contain certificates. The logic is bit more complicated. IIRC the certs are copied to operator filesystem in case of operator restart (or rescheduling to another node?)

Reply:
Can't they be mounted in the file system directly from the secrets, like it's done in the Kafka auto-provisioning? 

https://github.com/jaegertracing/jaeger-operator/blob/5999ea6f1057cc63d30888272e9f101a697da8f6/pkg/strategy/streaming.go#L145-L173

Reply:
They are mounted to Jaeger filesystem. But for the use case I am mentioned we need certs in the operator filesystem and not Jaeger instance (see the operator in my previous comment). The cert generation script then assures certs are valid and not expired...

Reply:
> But for the use case I am mentioned we need certs in the operator filesystem and not Jaeger instance (see the operator in my previous comment). The cert generation script then assures certs are valid and not expired...

This is another signal that there's a mismatch between things that should be operator-scoped and operand-scoped. Renewal of the certs, as well as creating/managing the secrets are operations that should be made by the operator independent of the reconciliation logic. The reconciliation logic should only be about the operand.

Reply:
> Renewal of the certs, as well as creating/managing the secrets are operations that should be made by the operator independent of the reconciliation logic.

Why handling objects/dependencie/whatever is required for the operand part of the reconciliation? My understanding of reconciliation logic is to prepare objects/dependencie/whatever for the operand.

Reply:
I don't know, you said it yourself, but I probably misunderstood what you meant...

> IIRC the certs are copied to operator filesystem in case of operator restart (or rescheduling to another node?)